include/aspect/fe_variable_system.h

```diff
@@ -0,0 +1,423 @@
+       * variables.
+       */
+      bool variable_exists(const std::string &name) const;
+
+      /**
+       * Return the list of all variables.
+       */
+      const std::vector<FEVariable<dim> > &get_variables() const;
+
+      /**
+       * Return the total number of components in the system.
+       */
+      unsigned int n_components() const;
+
+      /**
+       * Return the total number of block of the system.
+       */
+      unsigned int n_blocks() const;
+
+      /**
+       * Return the vector of finite element spaces used for the construction
+       * of the FESystem.
+       */
+      const std::vector<const FiniteElement<dim> *> &get_fes() const;
+
+      /**
+       * Return the vector of multiplicities used for the construction of the
+       * FESystem.
+       */
+      const std::vector<unsigned int> &get_multiplicities() const;
+
+      /**
+       * Return a variable that describes for each vector component which
+       * vector block it corresponds to.
+       */
+      const std::vector<unsigned int> &get_components_to_blocks() const;
+
+    protected:
+      std::vector<FEVariable<dim> > variables;
+
+      std::map<std::string, FEVariable<dim>* > name_map;
+      unsigned int n_components_;
+      unsigned int n_blocks_;
+      std::vector<const FiniteElement<dim> *> fes;
+      std::vector<unsigned int> multiplicities;
+      std::vector<unsigned int> components_to_blocks;
+  };
+
+
+
+
+  template <int dim>
+  VariableDefinition<dim>::VariableDefinition(const std::string &name,
+                                              std_cxx11::shared_ptr<FiniteElement<dim> > fe,
+                                              const unsigned int multiplicity,
+                                              const unsigned int n_blocks)
+    : name(name),
+      fe(fe),
+      multiplicity(multiplicity),
+      n_blocks(n_blocks)
+  {
+    Assert(n_blocks == 0
+           || n_blocks == 1
+           || n_blocks == n_components(),
+           ExcMessage("A Variable can only have 0, 1, or n_components() number of blocks"));
+  }
+
+  template <int dim>
+  unsigned int
+  VariableDefinition<dim>::n_components() const
+  {
+    return fe->n_components() * multiplicity;
+  }
+
+
+  template <int dim>
+  FEVariable<dim>::FEVariable(const VariableDefinition<dim> &fe_variable,
+                              const unsigned int component_index,
+                              const unsigned int block_index,
+                              const unsigned int base_index)
+    : VariableDefinition<dim> (fe_variable.name,
+                               fe_variable.fe,
+                               fe_variable.multiplicity,
+                               fe_variable.n_blocks),
+    first_component_index (component_index),
+    block_index (block_index),
+    base_index (base_index)
+  {
+    if (this->n_components()==1)
+      scalar_extractor.component = component_index;
```

Comment:
This seems fragile, since you assume that that's the name of the member variable -- which it is, but we don't ever seem to use it outside of `FEValues`. How about
```
  scalar_extractor = FEValuesExtractors::Scalar(component_index);
```
and below use a default constructed object for the else branch?